# Feedback from previous weeks and other and hints

1. Be careful about your environment remembering variables. Make sure your code works in a new _clean_ environment. In Colab: `Runtime`->`restart Runtime`, in Anaconda's Jupyter: `Kernel`->`Restart`.
2. Graphs without labels (or units when appropriate) are not worth any point.
3. Do put in sufficient explanatory comments in your code.
4. Functions are very important. Do look up the video on the Safari O'Reilly ressource if you are still not clear on them !

For this week you can use these imports at the start of your programs:

In [22]:
import numpy as np
import matplotlib.pyplot as plt

We will use a new module `uncertainties`, which is **not** standard in the colab environment. You will have to first run:

In [23]:
! pip install -q uncertainties

To install the module before you can import it:

In [24]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

# Introduction
In the practical classes PX2133/PX2233 and PX2338 (Obs tech), as well as your year 3/4 project, a lot of emphasis is placed on the determination and mathematical handling of errors.
The uncertainties module allows us to deal very easily with [error propagation](https://en.wikipedia.org/wiki/Propagation_of_uncertainty). For this sheet you should remind yourself about error bars in measurements and about propagation of uncertainties. Take an example from your lab handbook:

**Example 1**: If the length of a rectangle is $1.24\pm0.02 m$ and its breadth is $0.61\pm0.01 m$, what is its area and the error in the area? The following code snippet solves this problem in a few lines.

In [25]:
L = uc.ufloat(1.24, 0.02)
W = uc.ufloat(0.61, 0.01)
print ('Area is:', L*W, 'm^2') # Do remember to add the units when printing!

Area is: 0.756+/-0.017 m^2


**Note**: For the area itself, it's fairly straightforward:

In [26]:
1.24*0.61

0.7564

However, for the error bar on this number:

In [27]:
0.02*0.01

0.0002

does not work. Instead, the [error progation formula](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulae) gives:

In [28]:
np.abs(1.24*0.61)*np.sqrt((0.02/1.24)**2+(0.01/0.61)**2)

np.float64(0.01739540169125163)

So the area is $0.756\pm0.017 m^2$. `uncertainties` obviously saves a lot of work, even for such a simple case. You can also take a look at the web site uncertainties hosted at https://pythonhosted.org/uncertainties/user_guide.html. In particular, [this section](https://pythonhosted.org/uncertainties/user_guide.html#access-to-the-uncertainty-and-to-the-nominal-value) shows some of the properties of a `ufloat` you can access directly.

**Example 2**: A reference object is $10.0\pm0.0001 m$ long, and makes a viewing angle of $0.62\pm0.02 rad$. How far is it?

In [29]:
L = uc.ufloat(10.0, 0.0001)
theta = uc.ufloat(0.62,0.02)

Distance = (L/2)/um.tan(theta/2)

print ('Distance is:', Distance.nominal_value, 'm, with an error of:', Distance.std_dev)

Distance is: 15.609024890896208 m, with an error of: 0.537283338762715


Note the need to use "umath" functions (like `um.tan()` instead of `np.tan()`), and how to get the nominal value and the standard deviation of the uncertainties objects. To get nicer looking output, such as controlling the number of significant digits printed, you can use the information about formatting at https://docs.python.org/3/tutorial/inputoutput.html. In the exercises below you need to print the values to the screen. (Don’t forget units.)

# Exercises
This must be marked before you leave the lab. Mark weighting is in brackets.
**Save your work to GitHub after having run all cells with `Runtime` -> `Restart and run all`. And do not change the notebook's filename.** Do add comments to your code, you'll lose points if your code is hard to understand. Graphs without labels (or units when appropriate) are not worth any point.

## Exercise 0
[0] With some approximations, we have measured the mass of the following black-holes:
```
"35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"
```
Compute for each (with error-bars) their lifetime due to Hawking radiation:
$$
t = \left(\frac{M}{M_{\odot}}\right)^3\,\times\,2.097\,\times\,10^{67} yr
$$

(this exercise is for demonstration purposes and won't be marked)

In [30]:
! pip install -q uncertainties
import uncertainties as uc
import uncertainties.umath as um # for maths functions

def lifetime(mass):
  #This function will return the lifetime, in years, of a black hole of mass m in solar masses
  return (mass**3)*2.097*10e67

In [31]:
strings=["35.6+/-3.9","30.6+/-3.7","63.1+/-3.2","23.2+/-9.8","13.6+/-4.5","35.7+/-6.8","13.7+/-6.0","7.7+/-2.4","20.5+/-4.0"]

In [32]:
masses=[uc.ufloat_fromstr(s) for s in strings]

In [33]:
['{:10.1e}'.format(lifetime(bh_mass)) for bh_mass in masses]

['   9.5e+72+/-   3.1e+72',
 '   6.0e+72+/-   2.2e+72',
 '   5.3e+73+/-   0.8e+73',
 '   2.6e+72+/-   3.3e+72',
 '   5.3e+71+/-   5.2e+71',
 '   9.5e+72+/-   5.5e+72',
 '   5.4e+71+/-   7.1e+71',
 '   9.6e+70+/-   9.0e+70',
 '   1.8e+72+/-   1.1e+72']

## Exercise 1
[2] An object is measured to travel a distance $x = 5.1 \pm 0.4 m$ during a time of $t = 0.4 \pm 0.1 s$. What is the average velocity and the error in the average velocity?

In [10]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

#Convert the distance, x, and time, t, into strings with their uncertainties
xstring="5.1+/-0.4"
tstring="0.4+/- 0.1"

#Find the uncertainties of either string
x=uc.ufloat(5.1, 0.4)
t=uc.ufloat(0.4, 0.1)

#Define velocity as the equation x/t, as we have data for each variable
def velocity(x,t):
  #This function will return the average velocities, in ms-1, of an object by dividing distance, x, by time, t
  return x/t

#Format the output with one decimal place
['{:10.1e}'.format(velocity(x,t))]

['   1.3e+01+/-   0.3e+01']

## Exercise 2
[2] An enterprising cow attempts to jump over the moon by jumping vertically into the air with initial speed $v_0=4.0\pm0.2 m/s$. After a time $t=0.60\pm0.06s$, the height of the cow is $h = v_0t-\frac{1}{2}g t^2 = 0.636 m$. What is the uncertainty in $h$? Take $g$ as exactly $9.81 ms^{-2}$.

In [18]:
pwd

'C:\\Users\\c24061346'

In [11]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

#Convert the initial speed, vo, and time, t, into strings with their uncertainties
vostring="4.0+/-0.2"
tstring="0.60+/- 0.06"

#Find the uncertainties of either string
vo=uc.ufloat_fromstr(vostring)
t=uc.ufloat_fromstr(tstring)

#Define height as the equation above
def height(vo,t):
  #This function will return the height, in m, of the cow using the equation
  return (vo*t)-(0.5*9.81*(t)**2)

#Format the output with one decimal place
['{:10.1e}'.format(height(vo,t))]

['   6.3e-01+/-   1.6e-01']

## Exercise 3
[2] In an optics experiment the object distance $u$ is measured to be 20cm and the image distance $v$ is 10cm, both to an accuracy of 0.5cm. Find the focal length $f$ of the lens using the formula:

$$ \frac{1}{u}+\frac{1}{v}=\frac{1}{f}$$

In [12]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions

#Convert the object distance, u, and image distance, v, into strings with their uncertainties. In metres.
ustring="0.2+/-0.005"
vstring="0.1+/- 0.005"

#Find the uncertainties of either string
u=uc.ufloat_fromstr(ustring)
v=uc.ufloat_fromstr(vstring)

#Define focal length, f, as the equation above rearranged
def focallength(u,v):
  #This function will return the focal length, in m
  return (u*v)/(u+v)

#Format the output with one decimal place
['{:10.1e}'.format(focallength(u,v))]

['   6.7e-02+/-   0.2e-02']

## Exercise 4
[2] Two students each measure the refractive index of water. Jack measures a value of $1.33 \pm 0.03$ while Jill measures $1.28 \pm 0.02$. Are these values in agreement? *You do have to think a bit about this one...*

In [13]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions
import numpy as np

#Defining the measurements as floats
jackstring=uc.ufloat(1.33, 0.03)
jillstring=uc.ufloat(1.28, 0.02)

#Minus Jill's measurement from Jack's
print(jackstring - jillstring)

print("The result is not within the errors, therefore they don't agree")

0.05+/-0.04
The result is not within the errors, therefore they don't agree


## Exercise 5
[2] The damped resonance frequency $\omega_{res}$ of an oscillating system is related to the (un-damped) natural angular frequency $\omega_0$ and the damping coefficient $\alpha$ by:

$$\omega_{res} = \sqrt{ \omega_0^2 - 2\alpha^2}$$

Find $f_0$ if the measured resonance frequency $f_{res}$ is $23.2\pm0.1 Hz$ and the measured damping coefficient is $19.5\pm0.5s^{-1}$.


In [14]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions
import numpy as np

#Convert the resonance frequency, fres, and iamping coefficient, dc, into strings with their uncertainties
ufresstring="23.2+/-0.1"
dcstring="19.5+/- 0.5"

#Find the uncertainties of either string
ufres=uc.ufloat_fromstr(ufresstring)
dc=uc.ufloat_fromstr(dcstring)

#Using 𝜔o = 2*pi*fo, the above equation can be rearraanged to calculate fo
#Define fo, as the equation above rearranged
def frequencyo(ufres,dc):
  #This function will return fo, in Hz
  return um.sqrt(((ufres*2*np.pi)**2)+2*(dc**2)/2*np.pi)

#Format the output with one decimal place
['{:10.1e}'.format(frequencyo(ufres,dc))]

['   1.5e+02+/-   0.0e+02']

## Exercise 6
[2] Suppose you have the following equation from one of your lab experiments:

$$f=\frac{c}{2}\sqrt{\frac{n_x^2}{L_x^2}+\frac{n_y^2}{L_y^2}+\frac{n_z^2}{L_z^2}}$$

where $f$ is the resonant frequency of sound waves in a box of sides $L_x$, $L_y$ and $L_z$ in length and the $n_x$ etc. are integers. $L_x = 10.2\pm0.2m$, $L_y = 5.2\pm0.3m$ and $L_z = 20.0\pm0.1 m$, while $c = 331.3 + T * 0.606 \,m\,s^{-1}$ is the temperature-dependent speed of sound, and the temperature $T$ is $23 \pm 1^\circ C$.
Calculate $f$ and the error in $f$ for the following values of $(nx,ny,nz)= (1,1,1), (1,1,2)$ and $(2,1,1)$.


In [15]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions
import numpy as np

#Convert the sides of the box, Lx, Ly, Lz, and the temperature, T, into strings with their uncertainties
Lxstring="10.2+/-0.2"
Lystring="5.2+/-0.3"
Lzstring="20+/-0.1"
tstring="23+/-1"

#Find the uncertainties of all strings
Lx=uc.ufloat_fromstr(Lxstring)
Ly=uc.ufloat_fromstr(Lystring)
Lz=uc.ufloat_fromstr(Lzstring)
t=uc.ufloat_fromstr(tstring)

#Define c
c = 331.3 + t*0.606

#(𝑛𝑥,𝑛𝑦,𝑛𝑧)=(1,1,1)
#Define f for nx, as the equation above rearranged
def f111(Lx,Lny,Lnz,c):
  nx=1
  ny=1
  nz=1
  #This function will return fnx, in Hz
  return (c/2)*um.sqrt((nx**2/Lx**2)+(ny**2/Ly**2)+(nz**2/Lz**2))

#Format the output with one decimal place
#['{:10.1e}'.format(f111(Lx,Ly,Lz,c))]
print(f111(Lx,Ly,Lz,c))

#(𝑛𝑥,𝑛𝑦,𝑛𝑧)=(1,1,2)
#Define f for ny, as the equation above rearranged
def f112(Lx,Ly,Lz,c):
  nx=1
  ny=1
  nz=2
  #This function will return fny, in Hz
  return (c/2)*um.sqrt((nx**2/Lx**2)+(ny**2/Ly**2)+(nz**2/Lz**2))

#Format the output with one decimal place
#['{:10.1e}'.format(f112(Lx,Ly,Lz,c))]
print(f112(Lx,Ly,Lz,c)) 

#(𝑛𝑥,𝑛𝑦,𝑛𝑧)=(2,1,1)
#Define f for nz, as the equation above rearranged
def f211(Lx,Ly,Lz,c):
  nx=2
  ny=1
  nz=1
  #This function will return fnz, in Hz
  return (c/2)*um.sqrt((nx**2/Lx**2)+(ny**2/Ly**2)+(nz**2/Lz**2))

#Format the output with one decimal place
#['{:10.1e}'.format(f211(Lx,Ly,Lz,t,c))]
print(f211(Lx,Ly,Lz,c))

38.2+/-1.7
41.1+/-1.6
48.2+/-1.4


## Exercise 7
[4] The reflection coefficient $R_\parallel$ for parallel plane-polarised light reflected from a surface is given by the equation:

$$ R_\parallel = \frac{\tan^2(\theta_i - \theta_t)}{\tan^2(\theta_i + \theta_t)} $$

Calculate the error in $R_\parallel$ given measurements $\theta_i = (78 \pm 1)^\circ$ and $\theta_t = (40 \pm 1)^\circ$.

In [16]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions
import numpy as np

#Converting the measurements and errors to radians
i=78*(np.pi/180)
t=40*(np.pi/180)
error=1*(np.pi/180)

#Defining the measurements as floats
i=uc.ufloat(i, error)
t=uc.ufloat(t, error)

#Define R∥, as the equation above
def R(i,t):
  #This function will return fo, in Hz
  return ((um.tan(i-t)**2)/(um.tan(i+t)**2))

  #Format the output with one decimal place
['{:10.1e}'.format(R(i,t))]

['   1.7e-01+/-   0.3e-01']

## Exercise 8
[4] Calculate and print to the screen the fractional uncertainty, as a percentage to one
significant figure, of the fluid flow discharge coefficient $C_d$ from the equation

$$
C_d = \frac{\dot{m}\sqrt{1-\left(\frac{d}{D}\right)^4}}{Kd^2F\sqrt{\rho\Delta P}}
$$

where

\begin{align*}
    C_d &= \text{discharge coefficient}&& \text{(no units)} \\
    \dot{m} &= \text{mass flow rate}&& = 0.13 \pm 0.01kg\,s^{-1} \\
    d &= \text{orifice diameter}&& = 11\pm 1 mm \\
    D &= \text{pipe diameter}&& = 71 \pm 1 mm \\
    \rho &= \text{fluid density}&& =1.01\pm0.01g\,cm^{-3} \\
    \Delta P &= \text{differential pressure}&& =156 \pm 7 Pa \\
    K &= \text{a constant parameter}&& =\text{constant (no units)} \\
    F &= \text{thermal expansion factor}&& =\text{constant (no units)}
\end{align*}


In [17]:
import uncertainties as uc
import uncertainties.umath as um # for maths functions
import numpy as np

#Defining the measurements as floats
m=uc.ufloat(0.13,0.01)
d=uc.ufloat(0.011, 0.001)
D=uc.ufloat(0.071, 0.001)
rho=uc.ufloat(1.01,0.01)
deltaP=uc.ufloat(156,7)

#Setting to 1
K=1
F=1

def Cd(m,d,D,rho,deltaP):
  #This function will return Cd, in m^3/s
  return ((m*um.sqrt(1-(d*D)**4))/(K*d**2*F*um.sqrt(rho*deltaP)))

  #Calculating the fractional error
insidebracket=Cd(m,d,D,rho,deltaP)
a=insidebracket.std_dev/insidebracket.nominal_value

  #Format the output with one decimal place, with a percentage error
'{:.0%}'.format(a)


'20%'

## Exercise 9: Optional problem (not marked)
If you have time and want to try something interesting, do the following problem by plotting in 2D:
 - Draw an equilateral triangle with vertices and coordinates: vertex 1: $(p_1,q_1)$; vertex 2: $(p_2, q_2)$; vertex 3: $(p_3, q_3)$.
 - Place a dot at an arbitrary point $P = (x_0, y_0)$ within this triangle.
 - Find the next point by selecting randomly an integer $n = 1 , 2, $  or $3$ :
    1. If 1 , place a dot halfway between P and vertex 1.
    2. If 2 , place a dot halfway between P and vertex 2.
    3. If 3 , place a dot halfway between P and vertex 3.
 - Repeat the last two steps using the last dot as the new P.

Mathematically, the coordinates of successive points are given by the formulae

$$(x_{i+1},y_{i+1})=0.5[(x_i,y_i)+(p_n,q_n)]$$

and

$$n=int(1+3r_i),$$

where $r_i$ is a random number between 0 and 1 and where the $int()$ function outputs the closest integer smaller than or equal to the argument.

Try extending this to four vertices.